plotnine/utils.py

```diff
@@ -1041,7 +1068,10 @@ def order_as_data_mapping(*args):
     return data, mapping
 
 
-def is_data_like(obj):
+# Returning a type guard here is not fully sound, because if `obj`
+# is a callable, we aren't checking that it has no required args
+# and we can't check the return value's type.
+def is_data_like(obj: Any) -> TypeGuard[DataLike]:
```

Comment:
This is a little head spinner. At the moment I have no idea how to clear it up nicely. We can inspect, but that is overkill. Maybe remove it later!

Reply:
The type information is available in the user's code when they construct any of the various classes that accept data, it just gets lost along the way because both data and aesthetic mappings are accepted as positional params in any order and are then rearranged to fit.

The code could be rearranged so the type information is never lost.

Reply:
I've narrowed the input type from `Any` to `DataLike | aes | None`. In that case, if the input is a Callable, then it must have the right signature or a type error would be thrown in the user's code, if they use a type checker.

Even with that, there is still the problem that the different allowed input types like callables and objects with `to_pandas` methods are not mutually exclusive. If an object has a `to_pandas()` method with the right signature but is also callable with the wrong signature, then the type checker would accept it. When conversion is then attempted by calling the object first, that would fail and a runtime error would be thrown. The inverse problem could also happen if `to_pandas()` conversion were attempted first in the library.

There are a couple options:
1. Require the user to specify the conversion method
2. Try the other conversion methods, if one fails
3. Do nothing of the sort and just live with the corner case because objects that are both callable and have a to_pandas() method aren't common (no idea if this is true)

Reply:
Option 3, "Do nothing" is enough. To my knowledge `df.to_pandas()` only applies to polars objects. Also, the future is probably polars with first class support then pandas frames will be converted, but that is still further down the line.